# 1.  **Título del Tema**


**Sintaxis Fundamental y Tipos de Datos Esenciales**

# 2.  **Explicación Conceptual Detallada**


#### Sintaxis en Python

*   **Definición y Propósito:** La sintaxis de Python se refiere al conjunto de reglas que definen cómo se escribe un programa Python correctamente estructurado. El propósito es proporcionar una forma clara e inequívoca para que los programadores comuniquen sus intenciones al intérprete de Python (el programa que ejecuta tu código).

*   **Conceptos Clave Asociados:**
    *   **Palabras Clave (Keywords):** Palabras reservadas con significados especiales en Python (ej: `if`, `for`, `def`, `class`, `True`, `False`, `None`). No puedes usarlas como nombres de variables.
    *   **Identificadores:** Nombres que das a variables, funciones, clases, etc. Deben seguir ciertas reglas (empezar con letra o guion bajo, pueden contener letras, números y guion bajo). Python es sensible a mayúsculas y minúsculas (`miVariable` es diferente de `mivariable`).
    *   **Comentarios:** Texto que Python ignora, usado para explicar el código. Comienzan con `#` para una línea, o se usan triples comillas `""" ... """` o `''' ... '''` para multilínea (docstrings).
    *   **Indentación:** ¡Crucial en Python! Python usa espacios en blanco (generalmente 4 espacios por nivel) al inicio de una línea para definir bloques de código (por ejemplo, el cuerpo de una función, un bucle o una condición). No usa llaves `{}` como otros lenguajes para esto.
    *   **Declaraciones (Statements):** Instrucciones que Python puede ejecutar (ej: `x = 5`, `print("Hola")`).
    *   **Expresiones (Expressions):** Combinaciones de valores, variables y operadores que se evalúan para producir un valor (ej: `5 + 3`, `x * 2`).

*   **"Gotchas" o Errores Comunes:**
    *   `IndentationError`: Error más común para principiantes, debido a una indentación incorrecta o inconsistente.
    *   `SyntaxError`: Ocurre cuando el código no sigue las reglas gramaticales de Python.
    *   Olvidar los dos puntos `:` al final de declaraciones que inician un bloque (ej: `if`, `for`, `def`).
    *   Sensibilidad a mayúsculas/minúsculas: `variable` y `Variable` son diferentes.

#### Tipos de Datos en Python


*   **Definición y Propósito:** Los tipos de datos son clasificaciones que especifican qué tipo de valor tiene una variable y qué tipo de operaciones matemáticas, relacionales o lógicas se pueden aplicar sin causar un error. Python es un lenguaje de **tipado dinámico**, lo que significa que no necesitas declarar explícitamente el tipo de una variable; Python lo infiere en tiempo de ejecución.

*   **Tipos de Datos Fundamentales:**
    *   **Numéricos:**
        *   `int`: Números enteros (ej: `-5`, `0`, `100`).
        *   `float`: Números de punto flotante (decimales) (ej: `-3.14`, `0.0`, `99.99`).
        *   `complex`: Números complejos (ej: `3+5j`) (menos común para empezar).
    *   **Secuencias (ordenadas):**
        *   `str`: Cadenas de caracteres (texto). Se definen con comillas simples `'...'`, dobles `"..."` o triples `'''...'''` o `"""..."""`. Son **inmutables**.
        *   `list`: Colecciones ordenadas y **mutables** de ítems. Se definen con corchetes `[]` y los elementos separados por comas (ej: `[1, "hola", 3.0]`).
        *   `tuple`: Colecciones ordenadas e **inmutables** de ítems. Se definen con paréntesis `()` (ej: `(1, "hola", 3.0)`).
    *   **Mapeo (pares clave-valor):**
        *   `dict`: Diccionarios. Colecciones no ordenadas (aunque en Python 3.7+ mantienen el orden de inserción) de pares `clave:valor`. Son **mutables**. Se definen con llaves `{}` (ej: `{"nombre": "Ana", "edad": 30}`).
    *   **Conjuntos:**
        *   `set`: Colecciones desordenadas de ítems **únicos**. Son **mutables**. Se definen con llaves `{}` pero sin pares clave-valor (ej: `{1, 2, 3}`) o usando `set()`.
        *   `frozenset`: Versión **inmutable** de un `set`.
    *   **Booleanos:**
        *   `bool`: Representa valores de verdad. Solo puede ser `True` o `False`.
    *   **Tipo Nulo:**
        *   `NoneType`: Tiene un único valor: `None`. Se usa para representar la ausencia de un valor o un estado nulo.

*   **Importancia:** Conocer los tipos de datos es crucial porque determina:
    *   Qué operaciones puedes realizar (ej: no puedes sumar un número a una cadena de texto directamente sin convertir uno de ellos).
    *   Cuánta memoria ocupa una variable.
    *   Cómo se comportan los datos (mutabilidad vs. inmutabilidad).


*   **Mutabilidad e Inmutabilidad:**
    *   **Objetos Inmutables:** Una vez creados, su valor no puede cambiar. Si intentas "modificar" un objeto inmutable, en realidad estás creando un nuevo objeto. Ejemplos: `int`, `float`, `str`, `tuple`, `frozenset`.
    *   **Objetos Mutables:** Su valor puede cambiar después de su creación. Ejemplos: `list`, `dict`, `set`.


*   **Verificación y Conversión de Tipos:**
    *   `type()`: Función incorporada para obtener el tipo de un objeto.
    *   Conversión de tipos (Type Casting): Funciones como `int()`, `float()`, `str()`, `list()`, `tuple()`, `set()` permiten convertir un valor de un tipo a otro (si la conversión es válida).


*   **Ventajas de los Tipos de Datos en Python:**
    *   **Tipado Dinámico:** Facilita la escritura rápida de código y flexibilidad.
    *   **Riqueza de Tipos Incorporados:** Python ofrece una gran variedad de tipos de datos útiles listos para usar.


*   **Posibles Limitaciones o "Gotchas":**
    *   El tipado dinámico puede llevar a errores en tiempo de ejecución si no se manejan cuidadosamente los tipos (`TypeError`).
    *   La mutabilidad puede tener efectos secundarios inesperados si varias variables apuntan al mismo objeto mutable y una de ellas lo modifica.


*   **Buenas Prácticas:**
    *   Usa nombres de variables descriptivos.
    *   Entiende la mutabilidad de los objetos con los que trabajas.
    *   Verifica los tipos si recibes datos de fuentes externas o si la lógica de tu programa depende críticamente de ellos.
    *   Utiliza comentarios para aclarar el propósito de las variables y las secciones de código complejas.

# 3.  **Sintaxis y Ejemplos Básicos**


In [8]:
# Enteros y flotantes
edad = 25
precio = 19.99
print(f"edad: {edad} tipo: {type(edad)},\nprecio: {precio},  tipo: {type(precio)}")

edad: 25 tipo: <class 'int'>,
precio: 19.99,  tipo: <class 'float'>


In [11]:
# Cadenas
nombre = "Ana"
mensaje = 'Hola Mundo'
print(f"nombre: {nombre} tipo: {type(nombre)},\nmensaje: {mensaje},  tipo: {type(mensaje)}")

nombre: Ana tipo: <class 'str'>,
mensaje: Hola Mundo,  tipo: <class 'str'>


In [12]:
# Booleanos
es_valido = True
tiene_saldo = False
print(f"es_valido: {es_valido} tipo: {type(es_valido)},\ntiene_saldo: {tiene_saldo},  tipo: {type(tiene_saldo)}")

es_valido: True tipo: <class 'bool'>,
tiene_saldo: False,  tipo: <class 'bool'>


In [14]:
# Listas
frutas = ["manzana", "banana", "naranja"]
print(f"frutas: {frutas} tipo: {type(frutas)}")

frutas: ['manzana', 'banana', 'naranja'] tipo: <class 'list'>


In [15]:
# Tuplas
coordenadas = (40.7128, -74.0060)
print(f"coordenadas: {coordenadas} tipo: {type(coordenadas)}")

coordenadas: (40.7128, -74.006) tipo: <class 'tuple'>


In [16]:
# Diccionarios
usuario = {"nombre": "Carlos", "edad": 30}
print(f"usuario: {usuario} tipo: {type(usuario)}")

usuario: {'nombre': 'Carlos', 'edad': 30} tipo: <class 'dict'>


# 4.  **Documentación y Recursos Clave**


*   **Documentación Oficial de Python:**
    *   [Tipos de Datos Incorporados (Built-in Types)](https://docs.python.org/es/3/library/stdtypes.html)
    *   [Introducción Informal a Python (Cubre sintaxis básica)](https://docs.python.org/es/3/tutorial/introduction.html)


*   **Recursos Externos de Alta Calidad:**
    *   [Real Python - Basic Data Types in Python](https://realpython.com/python-data-types/): Un excelente artículo que profundiza en los tipos de datos.
    *   [W3Schools - Python Syntax](https://www.w3schools.com/python/python_syntax.asp): Referencia rápida y concisa sobre la sintaxis.

# 5.  **Ejemplos de Código Prácticos**


### Ejemplo 1: Explorando Tipos de Datos y Operaciones Básicas

In [ ]:
# Explorando Tipos de Datos y Operaciones Básicas

# ---- Variables Numéricas ----
edad = 30               # int
altura = 1.75           # float
saldo_bancario = -50.25 # float (puede ser negativo)

# Operaciones numéricas
suma_edades = edad + 5
nueva_altura = altura * 1.02 # Incremento del 2%
print(f"Edad en 5 años: {suma_edades}")
print(f"Nueva altura: {nueva_altura}")
print(f"Tipo de saldo_bancario: {type(saldo_bancario)}")
print("---")

# ---- Variables de Texto (Strings) ----
nombre = "Ana"
apellido = "Pérez"

# Concatenación de strings
nombre_completo = nombre + " " + apellido
print(f"Nombre completo: {nombre_completo}")

# Repetición de strings
saludo_repetido = "Hola! " * 3
print(f"Saludo repetido: {saludo_repetido}")

# Acceder a un carácter (los strings son secuencias)
primera_letra_nombre = nombre[0]
print(f"Primera letra de {nombre}: {primera_letra_nombre}")
print("---")

# ---- Variables Booleanas ----
es_mayor_de_edad = edad >= 18
tiene_buen_saldo = saldo_bancario > 0
print(f"¿Es Ana mayor de edad?: {es_mayor_de_edad}")
print(f"¿Tiene Ana buen saldo?: {tiene_buen_saldo}")
print(f"Tipo de es_mayor_de_edad: {type(es_mayor_de_edad)}")
print("---")

# ---- Tipo None ----
trabajo_actual = None # Ana actualmente no tiene trabajo
print(f"Trabajo actual de Ana: {trabajo_actual}")
print(f"Tipo de trabajo_actual: {type(trabajo_actual)}")

Edad en 5 años: 35
Nueva altura: 1.7850000000000001
Tipo de saldo_bancario: <class 'float'>
---
Nombre completo: Ana Pérez
Saludo repetido: Hola! Hola! Hola! 
Primera letra de Ana: A
---
¿Es Ana mayor de edad?: True
¿Tiene Ana buen saldo?: False
Tipo de es_mayor_de_edad: <class 'bool'>
---
Trabajo actual de Ana: None
Tipo de trabajo_actual: <class 'NoneType'>


### Ejemplo 2: Trabajando con Listas y Diccionarios

In [22]:
# Celda 2: Trabajando con Listas y Diccionarios

# ---- Listas (mutables, ordenadas) ----
compras_supermercado = ["manzanas", "leche", "pan", "huevos"]
print(f"Lista de compras inicial: {compras_supermercado}")

# Acceder a un elemento (índices empiezan en 0)
primer_item = compras_supermercado[0]
print(f"Primer ítem: {primer_item}")

# Modificar un elemento
compras_supermercado[1] = "leche deslactosada"
print(f"Lista de compras modificada: {compras_supermercado}")

# Añadir un elemento al final
compras_supermercado.append("yogur")
print(f"Lista con nuevo ítem: {compras_supermercado}")

# Longitud de la lista
num_items = len(compras_supermercado)
print(f"Número de ítems en la lista: {num_items}")
print("---")

# ---- Diccionarios (mutables, pares clave-valor) ----
informacion_persona = {
    "nombre": "Carlos",
    "edad": 28,
    "ciudad": "Madrid",
    "hobbies": ["leer", "correr", "cocinar"]
}
print(f"Información inicial de la persona: {informacion_persona}")

# Acceder a un valor usando su clave
nombre_persona = informacion_persona["nombre"]

edad_persona = informacion_persona.get("edad") # .get() es más seguro si la clave podría no existir
print(f"Nombre: {nombre_persona}, Edad: {edad_persona}")

apellido_persona = informacion_persona.get("apellido") # .get() es más seguro si la clave podría no existir
print(f"Nombre: {nombre_persona}, Apellido: {apellido_persona} con .get()")

# apellido_persona = informacion_persona["apellido"] genera error
# print(f"Nombre: {nombre_persona}, Apellido: {apellido_persona} sin .get()") genera error

# Modificar un valor
informacion_persona["ciudad"] = "Barcelona"
print(f"Información con ciudad actualizada: {informacion_persona}")

# Añadir un nuevo par clave-valor
informacion_persona["profesion"] = "Ingeniero"
print(f"Información con profesión: {informacion_persona}")

# Acceder a una lista dentro del diccionario
hobbies_carlos = informacion_persona["hobbies"]
print(f"Hobbies de Carlos: {hobbies_carlos}")
print(f"Primer hobbie de Carlos: {hobbies_carlos[0]}")

Lista de compras inicial: ['manzanas', 'leche', 'pan', 'huevos']
Primer ítem: manzanas
Lista de compras modificada: ['manzanas', 'leche deslactosada', 'pan', 'huevos']
Lista con nuevo ítem: ['manzanas', 'leche deslactosada', 'pan', 'huevos', 'yogur']
Número de ítems en la lista: 5
---
Información inicial de la persona: {'nombre': 'Carlos', 'edad': 28, 'ciudad': 'Madrid', 'hobbies': ['leer', 'correr', 'cocinar']}
Nombre: Carlos, Edad: 28
Nombre: Carlos, Apellido: None con .get()
Información con ciudad actualizada: {'nombre': 'Carlos', 'edad': 28, 'ciudad': 'Barcelona', 'hobbies': ['leer', 'correr', 'cocinar']}
Información con profesión: {'nombre': 'Carlos', 'edad': 28, 'ciudad': 'Barcelona', 'hobbies': ['leer', 'correr', 'cocinar'], 'profesion': 'Ingeniero'}
Hobbies de Carlos: ['leer', 'correr', 'cocinar']
Primer hobbie de Carlos: leer


### Ejemplo 3: Conversión de Tipos e Inmutabilidad de Strings y Tuplas

In [23]:
# Celda 3: Conversión de Tipos e Inmutabilidad

# ---- Conversión de Tipos ----
puntos_str = "150"
puntos_int = int(puntos_str)  # Convertir string a entero
print(f"Puntos (str): '{puntos_str}', Tipo: {type(puntos_str)}")
print(f"Puntos (int): {puntos_int}, Tipo: {type(puntos_int)}")

puntos_adicionales = 25.5
total_puntos_float = puntos_int + puntos_adicionales # int + float = float
print(f"Total puntos (float): {total_puntos_float}, Tipo: {type(total_puntos_float)}")

# Convertir a string para concatenar
mensaje_puntos = "Total de puntos acumulados: " + str(total_puntos_float)
print(mensaje_puntos)
print("---")

# ---- Inmutabilidad de Strings ----
mi_string = "Hola Mundo"
print(f"String original: {mi_string}")
# mi_string[0] = "h" # Esto generaría un TypeError: 'str' object does not support item assignment
# Para "modificar" un string, creamos uno nuevo
mi_string_modificado = "h" + mi_string[1:]
print(f"String 'modificado' (en realidad nuevo): {mi_string_modificado}")
print("---")

# ---- Inmutabilidad de Tuplas ----
coordenadas = (10, 20, 30)
print(f"Tupla original: {coordenadas}")
print(f"Primer elemento de la tupla: {coordenadas[0]}")
# coordenadas[0] = 5 # Esto generaría un TypeError: 'tuple' object does not support item assignment
# Si necesitas modificarla, tendrías que convertirla a lista, modificar y luego (opcionalmente) a tupla de nuevo
coordenadas_lista = list(coordenadas)
coordenadas_lista[0] = 5
coordenadas_modificada_tupla = tuple(coordenadas_lista)
print(f"Tupla 'modificada' (pasando por lista): {coordenadas_modificada_tupla}")

Puntos (str): '150', Tipo: <class 'str'>
Puntos (int): 150, Tipo: <class 'int'>
Total puntos (float): 175.5, Tipo: <class 'float'>
Total de puntos acumulados: 175.5
---
String original: Hola Mundo
String 'modificado' (en realidad nuevo): hola Mundo
---
Tupla original: (10, 20, 30)
Primer elemento de la tupla: 10
Tupla 'modificada' (pasando por lista): (5, 20, 30)


# 6.  **Ejercicio Práctico**


**Título del Ejercicio:** Mini Perfil de Usuario

**Descripción:**
Vas a crear un pequeño script que almacene y muestre información básica sobre un usuario ficticio.

1.  **Crea las siguientes variables** con los datos que elijas:
    *   `nombre_usuario` (string)
    *   `edad_usuario` (integer)
    *   `altura_cm_usuario` (integer, ej: 175 para 1.75m)
    *   `email_usuario` (string)
    *   `es_estudiante_activo` (boolean)
    *   `cursos_inscrito` (list de strings, con al menos 2 nombres de cursos)

2.  **Convierte la altura de cm a metros** (float) y guárdala en una nueva variable llamada `altura_metros_usuario`. (Pista: 1 metro = 100 cm).

3.  **Imprime la información del usuario** de una manera legible y amigable. Deberías mostrar:
    *   El nombre del usuario.
    *   Su edad.
    *   Su altura en metros (con dos decimales).
    *   Su email.
    *   Si es un estudiante activo o no.
    *   El primer curso al que está inscrito.
    *   La cantidad total de cursos a los que está inscrito.

4.  **Verifica e imprime el tipo de dato** de al menos 3 de tus variables (ej: `nombre_usuario`, `altura_metros_usuario`, `cursos_inscrito`).

In [30]:
nombre_usuario = str
edad_usuario = int
altura_cm_usuario = int
email_usuario = str
es_estudiante_activo = bool
cursos_inscritos = list

print(f"nombre_usuario = {nombre_usuario} ,type: {type(nombre_usuario)}")
print(f"edad_usuario = {edad_usuario} ,type: {type(edad_usuario)}")
print(f"altura_cm_usuario = {altura_cm_usuario} ,type: {type(altura_cm_usuario)}")
print(f"email_usuario = {email_usuario} ,type: {type(email_usuario)}")
print(f"es_estudiante_activo = {es_estudiante_activo} ,type: {type(es_estudiante_activo)}")
print(f"cursos_inscritos = {cursos_inscritos} ,type: {type(cursos_inscritos)}")

nombre_usuario = <class 'str'> ,type: <class 'type'>
edad_usuario = <class 'int'> ,type: <class 'type'>
altura_cm_usuario = <class 'int'> ,type: <class 'type'>
email_usuario = <class 'str'> ,type: <class 'type'>
es_estudiante_activo = <class 'bool'> ,type: <class 'type'>
cursos_inscritos = <class 'list'> ,type: <class 'type'>


In [36]:
nombre_usuario = "nicolas"
edad_usuario = 20
altura_cm_usuario = 170
email_usuario = "nicolas@email"
es_estudiante_activo = True
cursos_inscritos = ["python","sql"]
altura_m_usuairo = altura_cm_usuario/100
total_cursos = len(cursos_inscritos)

print(f"nombre_usuario = {nombre_usuario}") 
print(f"edad_usuario = {edad_usuario}") 
print(f"altura_cm_usuario = {altura_cm_usuario}cm") 
print(f"email_usuario = {email_usuario}") 
print(f"es_estudiante_activo = {es_estudiante_activo}") 
print(f"cursos_inscritos = {cursos_inscritos}") 
print(f"altura_m_usuairo = {altura_m_usuairo}m") 
print(f"total_cursos = {total_cursos}") 

nombre_usuario = nicolas
edad_usuario = 20
altura_cm_usuario = 170cm
email_usuario = nicolas@email
es_estudiante_activo = True
cursos_inscritos = ['python', 'sql']
altura_m_usuairo = 1.7m
total_cursos = 2


# 7.  **Conexión con Otros Temas**


*   **Conceptos que debería conocer previamente:**
    *   Realmente, este es el punto de partida. Una idea general de qué es la programación puede ser útil, pero no se requieren conocimientos previos de Python.

*   **Temas futuros para los que este conocimiento será importante:**
    *   **Absolutamente TODO en Python.**
    *   **Operadores:** Aprenderás sobre los operadores aritméticos (`+`, `-`, `*`, `/`), de comparación (`==`, `!=`, `>`, `<`), lógicos (`and`, `or`, `not`), etc., que actúan sobre estos tipos de datos.
    *   **Estructuras de Control:** `if/elif/else` (condicionales) y `for/while` (bucles) toman decisiones y repiten acciones basadas en valores y tipos de datos (especialmente booleanos y secuencias).
    *   **Funciones:** Recibirán argumentos (que tienen tipos) y devolverán valores (que también tienen tipos).
    *   **Estructuras de Datos Avanzadas:** Listas, tuplas, diccionarios y sets son la base, pero construirás estructuras más complejas a partir de ellos.
    *   **Programación Orientada a Objetos (POO):** Crearás tus propios tipos de datos (clases) que encapsulan datos y comportamientos.
    *   **Manejo de Archivos:** Leerás y escribirás datos (que serán strings, números, etc.) desde y hacia archivos.
    *   **Librerías y Módulos:** Todas las librerías de Python (NumPy, Pandas, Django, etc.) manipulan estos tipos de datos fundamentales.

# 8.  **Aplicaciones en el Mundo Real**


*   **Desarrollo Web:**
    *   Almacenar información de usuarios (nombres - `str`, edades - `int`, emails - `str`, perfiles - `dict`).
    *   Gestionar el contenido de un blog (títulos - `str`, artículos - `str`, fechas - `datetime` (un tipo más avanzado), etiquetas - `list`).



*   **Análisis de Datos y Ciencia de Datos:**
    *   Procesar conjuntos de datos numéricos (ventas - `float`, conteos - `int`).
    *   Manipular datos categóricos (nombres de productos - `str`, regiones - `str`).
    *   Resultados de experimentos (mediciones - `float`, observaciones - `list` de `dict`).


*   **Automatización de Tareas:**
    *   Leer nombres de archivos (`str`), tamaños de archivos (`int`).
    *   Configuraciones de scripts (parámetros como `True`/`False` - `bool`, umbrales numéricos - `int`/`float`).


*   **Juegos:**
    *   Puntuaciones de jugadores (`int`).
    *   Nombres de personajes (`str`).
    *   Inventarios (`list` o `dict`).
    *   Posiciones en un mapa (coordenadas en `tuple` o `list` de `float`).